src/error.rs

```diff
@@ -44,6 +56,40 @@ impl<T> OptionalExtension<T> for Result<T, ConnectionError> {
     }
 }
 
+impl<T> OptionalExtension<T> for Result<T, RunError<DieselError>> {
+    fn optional(self) -> Result<Option<T>, ConnectionError> {
+        let self_as_query_result: diesel::QueryResult<T> = match self {
+            Ok(value) => Ok(value),
+            Err(RunError::User(error_kind)) => Err(error_kind),
+            Err(RunError::RuntimeShutdown) => return Err(ConnectionError::RuntimeShutdown),
+        };
+
+        self_as_query_result
+            .optional()
+            .map_err(ConnectionError::Query)
+    }
+}
+
+/// Errors encountered while running a function on a connection pool.
+#[derive(Error, Debug, Clone, Copy, Eq, PartialEq)]
+pub enum RunError<E> {
```

Comment:
Does this actually need to be generic? Could we just have `E` be `DieselError`?

I'm aware that (before this PR) there were spots where we used `E: From<DieselError>` -- seems like those spots could become `E: From<RunError>` instead, but without making `RunError` itself be generic.

Reply:
I'm not sure! I'll try it out. I was trying to minimize changes in the PR but not having a generic here would certainly simplify things.

Reply:
Removing the type parameter is possible as long as we change the `run` function to always return a `DieselError`. Another option is to introduce a separate error type just for that function.